pkg/flexvolume/gcepd/flexvolume.go

```diff
@@ -0,0 +1,283 @@
+	// "{\"status\": \"Success\", \"capabilities\": {\"attach\": false}}"
+	return map[string]interface{}{
+		"capabilities": map[string]bool{
+			"attach": false,
+		},
+	}, nil
+}
+
+// initFlexVolumeDriverForMount parse user provided jsonOptions to initialize FlexVolumeDriver.
+func (d *FlexVolumeDriver) initFlexVolumeDriverForMount(jsonOptions string) error {
+	var volOptions map[string]interface{}
+	json.Unmarshal([]byte(jsonOptions), &volOptions)
+
+	// TODO(harry): check "volId zone project" are not nil to avoid panic, below check is useless.
+	if len(volOptions[knownflags.VolIdKey].(string)) == 0 {
+		return fmt.Errorf("jsonOptions is not set by user properly: %#v", jsonOptions)
+	}
+
+	d.volId = volOptions[knownflags.VolIdKey].(string)
+	d.fsType = volOptions[knownflags.SystemFsTypeKey].(string)
+	d.zone = volOptions[knownflags.ZoneKey].(string)
+	d.project = volOptions[knownflags.ProjectKey].(string)
+
+	return nil
+}
+
+// initFlexVolumeDriverForUnMount use targetMountDir to initialize FlexVolumeDriver from tag file.
+func (d *FlexVolumeDriver) initFlexVolumeDriverForUnMount(targetMountDir string) error {
+	// Use the tag file to store volId since flexvolume will execute fresh new binary every time.
+	optsData, err := utilmetadata.ReadJsonOptsFile(targetMountDir)
+	if err != nil {
+		return err
+	}
+
+	d.volId = optsData[knownflags.VolIdKey].(string)
+	d.zone = optsData[knownflags.ZoneKey].(string)
+	d.project = optsData[knownflags.ProjectKey].(string)
+
+	return nil
+}
+
+// Invocation: <driver executable> attach <json options> <node name>
+func (d *FlexVolumeDriver) attach(jsonOptions, nodeName string) (map[string]interface{}, error) {
+	return nil, nil
+}
+
+// Invocation: <driver executable> detach <mount device> <node name>
+func (d *FlexVolumeDriver) detach(mountDev, nodeName string) (map[string]interface{}, error) {
+	return nil, nil
+}
+
+// Invocation: <driver executable> waitforattach <mount device> <json options>
+func (d *FlexVolumeDriver) waitForAttach(mountDev, jsonOptions string) (map[string]interface{}, error) {
+	return map[string]interface{}{"device": mountDev}, nil
+}
+
+// Invocation: <driver executable> isattached <json options> <node name>
+func (d *FlexVolumeDriver) isAttached(jsonOptions, nodeName string) (map[string]interface{}, error) {
+	return map[string]interface{}{"attached": true}, nil
+}
+
+// Invocation: <driver executable> mount <mount dir> <json options>
+// mount persist meta data generated from jsonOptions into a tag file in target dir.
+func (d *FlexVolumeDriver) mount(targetMountDir, jsonOptions string) (map[string]interface{}, error) {
+	if err := d.initFlexVolumeDriverForMount(jsonOptions); err != nil {
+		return nil, err
+	}
+
+	// Step 1: Attach GCE PD to the instance.
+	if err := attachDisk(d.project, d.zone, d.volId); err != nil {
+		return nil, err
+	}
+
+	// Step 2: Format the device.
+	if err := fomratDisk(d.volId, d.fsType); err != nil {
```

Comment:
@gnawux when `attachDisk ` is called, GCE need 2s finish attach, so this `formatDisk` may fail for once. Will not break existing test/demo since it will be recovered in seconds.

I am adding operation check here.

Reply:
2 seconds... ok, reasonable...